pkg/vsphere/performance/host_metrics.go

```diff
@@ -0,0 +1,178 @@
+type HostMemory struct {
+	ActiveKB   int64
+	ConsumedKB int64
+	OverheadKB int64
+	TotalKB    int64
+}
+
+// HostCPU stores an ESXi host's CPU metrics.
+type HostCPU struct {
+	UsagePercent float64
+}
+
+// HostMetricsInfo stores an ESXi host's memory and CPU metrics.
+type HostMetricsInfo struct {
+	Memory HostMemory
+	CPU    HostCPU
+}
+
+// HostMetrics returns CPU and memory metrics for all ESXi hosts in the input session's cluster.
+func HostMetrics(session *session.Session) (map[*object.HostSystem]*HostMetricsInfo, error) {
+	ctx := context.Background()
+	op := trace.NewOperation(ctx, "ESX host metrics collector")
+
+	if session == nil {
+		return nil, fmt.Errorf("session not set")
+	}
+
+	// Gather hosts from the session cluster and then obtain their morefs.
+	hosts, err := gatherHosts(op.Context, session)
+	if err != nil {
+		return nil, fmt.Errorf("unable to obtain host morefs from session: %s", err)
+	}
+	morefToHost := make(map[types.ManagedObjectReference]*object.HostSystem)
+	morefs := make([]types.ManagedObjectReference, len(hosts))
+	for i := range hosts {
+		moref := hosts[i].Reference()
+		morefToHost[moref] = hosts[i]
+		morefs[i] = moref
+	}
+
+	// Query CPU and memory metrics for the morefs.
+	spec := types.PerfQuerySpec{
+		Format:     string(types.PerfFormatNormal),
+		IntervalId: sampleInterval,
+	}
+
+	counters := []string{cpuUsage, memActive, memConsumed, memTotalCapacity, memOverhead}
+	perfMgr := performance.NewManager(session.Vim25())
+	sample, err := perfMgr.SampleByName(op.Context, spec, counters, morefs)
+	if err != nil {
+		errStr := "unable to get metric sample: %s"
+		op.Errorf(errStr, err)
+		return nil, fmt.Errorf(errStr, err)
+	}
+
+	results, err := perfMgr.ToMetricSeries(op.Context, sample)
+	if err != nil {
+		errStr := "unable to convert metric sample to metric series: %s"
+		op.Errorf(errStr, err)
+		return nil, fmt.Errorf(errStr, err)
+	}
+
+	metrics := assembleMetrics(op, morefToHost, results)
+	return metrics, nil
+}
+
+// assembleMetrics processes the metric samples received from govmomi and returns a finalized metrics map
+// keyed by the hosts.
+func assembleMetrics(op trace.Operation, morefToHost map[types.ManagedObjectReference]*object.HostSystem,
+	results []performance.EntityMetric) map[*object.HostSystem]*HostMetricsInfo {
+	metrics := make(map[*object.HostSystem]*HostMetricsInfo)
+
+	for _, host := range morefToHost {
+		metrics[host] = &HostMetricsInfo{}
+	}
+
+	for i := range results {
+		res := results[i]
+		host, exists := morefToHost[res.Entity]
+		if !exists {
+			op.Warnf("moref %s does not exist in requested morefs, skipping", res.Entity.String())
+			continue
+		}
+
+		// Process each value and assign it directly to the corresponding metric
+		// field since there is only one sample.
+		for _, v := range res.Value {
+
+			// We don't need to collect non-aggregate (non-empty Instance) metrics.
+			if v.Instance != "" {
+				continue
+			}
+
+			switch v.Name {
```

Comment:
I would check v.Value for the length, just in case. since accessing it by [0] is kinda spooky.

Reply:
Sounds good - will do.

Reply:
The i from `for i := range results` could be used in place of the zero

Reply:
Not sure if the index from `range results` will work. Each `result` holds a slice of metrics for a particular moref, and the metrics slice has a `Values` field with only one value since we're fetching one sample.

I've added the length check though.